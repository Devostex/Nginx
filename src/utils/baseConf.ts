import ConfigParser, { NginxConfig } from "@webantic/nginx-config-parser";
import { readFile } from "fs-extra";
import { join } from "path";

const parser = new ConfigParser();

console.log(
	parser.toJSON(`server {
    listen 443 ssl;
    server_name adguard.samtaen.nl;


    location / {
        proxy_pass http://adguard:80;
        sub_filter '</head>' '<link rel="stylesheet" type="text/css" href="/styles/adguard.css"></head>';
        sub_filter_once on;

    }
    
    location /styles {
        alias /www/files/styles;
    }
}
`)
);

const baseConf = async (): Promise<NginxConfig> => {
	const NginxConfig = (
		await readFile(join(process.cwd(), "src", "nginx", "baseConfig.conf"))
	).toString();
	return parser.toJSON(NginxConfig);
};
export default baseConf;
