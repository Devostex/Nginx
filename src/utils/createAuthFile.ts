import md5 from "apache-md5";
import { outputFile, pathExists } from "fs-extra";
import { join } from "path";

import createHash from "@utils/createHash";
import { authPath } from "@utils/env";

import { Auth } from "@models/ParsedConfig";

const htpasswd = ({ username, password }: Auth): string =>
	`${username}:${md5(password)}`;

const createAuthFile = async (
	users: Auth[]
): Promise<{ filename: string; hash: string }> => {
	const hash = createHash(JSON.stringify(users));
	const filename = join(authPath, hash);

	if (!(await pathExists(filename))) {
		const content = users.map(htpasswd).join("\n");

		await outputFile(filename, content);
	}

	return { filename, hash };
};

export default createAuthFile;
