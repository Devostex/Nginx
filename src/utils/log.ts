/* eslint-disable @typescript-eslint/explicit-function-return-type */
import chalk from "chalk";
import { YAMLException } from "js-yaml";
import { ZodIssue } from "zod";

import settings from "@utils/settings";

class Log {
	public log;

	constructor() {
		this.log = (...msgs: string[]) =>
			console.log(chalk`[{blue NCM}] ${msgs.join(" ")}`);
	}

	// Helpers

	private info(msg: string) {
		this.log(chalk`[{yellow INFO}] ${msg}`);
	}

	private done(msg: string) {
		this.log(chalk`[{green DONE}] ${msg}`);
	}

	private error(msg: string, ...args: string[]) {
		this.log(chalk`[{red ERROR}] ${msg}`, ...args);
	}

	private warn(msg: string) {
		this.log(chalk`[{red WARN}] ${msg}`);
	}

	// Global

	public started() {
		this.info(chalk`{green Started}`);
	}

	public finished(started: number) {
		this.done(chalk`{green Done in ${(Date.now() - started) / 1000}s}`);
	}

	public exited() {
		this.error(chalk`{red Exited}`);
	}

	public exception() {
		this.error("An error occurred:");
	}

	public noCertbotEmail() {
		this.error(
			chalk`{red You must set the {reset {bold {dim CERBOT_EMAIL}}} environment variable, certbot can't launch without it.}`
		);
	}

	// Old configurations

	public rmOld() {
		this.info(
			chalk`{yellow Removing old generated configs...} {dim ${settings.nginxConfigPath}}`
		);
	}

	public noOld() {
		this.error(
			chalk`{red Nginx Config Path not found:} {dim ${settings.nginxConfigPath}}`
		);
		this.info(
			chalk`You can set {dim NGINX_CONFIG_PATH} env variable to customize this location.`
		);
	}

	public nginxConfNotFound(nginxPath: string) {
		this.warn(chalk`nginx.conf does not exist: {dim ${nginxPath}}`);
	}
	// Nginx Config

	public configDone(config: string) {
		this.done(config);
	}

	// User Config

	public configNotFound(path: string) {
		this.error(chalk`{red There is no config file in: } {dim ${path}}`);
	}

	public configFolderNotFound(path: string) {
		this.error(
			chalk`{red The config {bold folder} was not found} {dim ${path}}`
		);
	}

	public multipleConfigs(configs: string[]) {
		this.warn(
			chalk`Multiple config files were found, using {yellow ${configs[0]}}`
		);
	}

	public invalidConfig(multiple: boolean) {
		this.error(
			chalk`There was an issue with your config, the error${
				multiple ? "s" : ""
			} ${multiple ? "are" : "is"} listed below.`
		);
	}

	public configError(config: string) {
		this.error(
			chalk`There was an error parsing the config, please check {dim ${config}}`
		);
	}

	public configYamlError(error: YAMLException) {
		this.error(
			chalk`Yaml Error: {dim ${error.reason}}:\n${error.mark.snippet}`
		);
	}

	public configJSONError(error: Error) {
		this.error(chalk`JSON Error: {dim ${error.message}}`);
	}

	public configJSError(error: Error) {
		this.error(
			chalk`JS Error: {dim ${error.message}}\n${
				error.stack?.split("\n")[1]
			}`
		);
	}

	public configEmpty() {
		this.error(chalk`The config is empty`);
	}

	public configIssue(issue: ZodIssue) {
		const path = issue.path.join(".") || "config";
		this.log();
		this.error(chalk`{magenta Path}: {dim ${path}}`);
		this.error(chalk`{magenta Error}: ${issue.message}`);
	}

	public configValid(file: string) {
		this.info(chalk`Config is valid {dim ${file}}`);
	}

	public configENVNotFound(envKey: string, env: string) {
		this.warn(
			chalk`Config contained {dim ${envKey}}, however {dim process.env.${env}} was not defined.`
		);
	}

	// CSS

	public downloadCSS(url: string) {
		this.info(chalk`{yellow Downloading CSS file...} {dim ${url}}`);
	}

	public cachedCSS(url: string) {
		this.done(chalk`{blue CSS file is cached}, skipping: {dim ${url}}`);
	}

	public CSSDownloaded(url: string) {
		this.done(chalk`{blue Downloaded CSS file} {dim ${url}}`);
	}

	public CSSError(url: string, error: string) {
		this.error(
			chalk`{red There was an error downloading/compressing the url:} {dim ${url}}`
		);
		this.error(chalk`{red The error: } ${error}`);
	}

	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	public CSSWriteError(filename: string, error: any) {
		this.error(
			chalk`{red There was an error writing to} {dim ${filename}}`
		);
		this.error(chalk`{red The error: }`, error?.message ?? error);
	}

	// JS

	public downloadJS(url: string) {
		this.info(chalk`{yellow Downloading JS file...} {dim ${url}}`);
	}

	public cachedJS(url: string) {
		this.done(
			chalk`{blue JS file is already downloaded, skipping:} {dim ${url}}`
		);
	}

	public JSDownloaded(url: string) {
		this.done(chalk`{blue Downloaded JS file} {dim ${url}}`);
	}

	// Cloudflare

	private cloudflareExpiryDays = (
		settings.cloudflareExpiry /
		1000 /
		60 /
		60 /
		24
	).toFixed(0);

	public updatingCloudflare() {
		this.info(chalk`{yellow Updating Cloudflare ips...}`);
	}

	public cloudflareCached() {
		this.info(
			chalk`{blue Not updating ips}, Current config has been updated less than {dim ${this.cloudflareExpiryDays}} days`
		);
	}

	public cloudflareExpired() {
		this.info(
			chalk`Cloudflare cached ips expired, Cache duration: {dim ${this.cloudflareExpiryDays}} days`
		);
	}

	public cloudflareUpdated(took: number) {
		this.done(
			chalk`Updated Cloudflare ips {dim Request took {yellow ${took}ms}}`
		);
	}

	public cloudflareUnchanged(took: number) {
		this.info(
			chalk`Cloudflare ips refreshed, no changes with cache {dim Request took {yellow ${took}ms}}`
		);
	}

	public cloudflareRealIp() {
		this.done(chalk`Cloudflare Real Ip List has been generated`);
	}
}

export default new Log();
