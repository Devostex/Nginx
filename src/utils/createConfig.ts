import ConfigParser, { NginxConfig } from "@webantic/nginx-config-parser";
import { join } from "path";

import baseConf from "@utils/baseConf";
import createHash from "@utils/createHash";
import downloadCSSToFile from "@utils/downloadCSSToFile";
import downloadJSToFile from "@utils/downloadJSToFile";
import env from "@utils/env";

import { Location, SimpleServer } from "@models/ParsedConfig";

const parser = new ConfigParser();

const createLocation = async (
	JsonConf: NginxConfig,
	location: Location
): Promise<void> => {
	const locString = `location ${location.location}`;
	const block = JsonConf.server[locString] ?? {};

	// Proxy Pass
	block.proxy_pass = location.proxy_pass;

	// Websockets
	if (location.websocket) {
		block.proxy_set_header ??= [];
		block.proxy_set_header.push(
			"Upgrade $http_upgrade",
			"Connection $http_connection"
		);

		block.proxy_http_version = 1.1;
	}

	// Custom CSS
	if (location.custom_css.length) {
		const fileNames = location.custom_css.map((g) => createHash(g));

		block.sub_filter = `'</head>' '${fileNames
			.map(
				(hash) =>
					`<link rel="stylesheet" type="text/css" href="/custom_assets/css/${hash}.css">`
			)
			.join("")}</head>'`;

		JsonConf.server["location /custom_assets"] = {
			alias: env.customFilesPath
		};

		await downloadCSSToFile(location.custom_css);
	}

	// Custom JS
	if (location.custom_js.length) {
		const fileNames = location.custom_js.map((g) => createHash(g));

		block.sub_filter = `'</body>' '${fileNames
			.map(
				(hash) => `<script src="/custom_assets/js/${hash}.js"></script>`
			)
			.join("")}</body>'`;

		JsonConf.server["location /custom_assets"] = {
			alias: env.customFilesPath
		};

		await downloadJSToFile(location.custom_js);
	}
	JsonConf.server[locString] = block;
};

const createConfig = async (server: SimpleServer): Promise<string> => {
	const JsonConf: NginxConfig = await baseConf();

	// Server Name
	JsonConf.server.server_name = server.server_name;

	// SSL Certificate files
	const sslKeysPath = join("/etc/letsencrypt/live", server.server_name, "/");
	JsonConf.server.ssl_certificate = sslKeysPath + "fullchain.pem";
	JsonConf.server.ssl_certificate_key = sslKeysPath + "privkey.pem";
	JsonConf.server.ssl_trusted_certificate = sslKeysPath + "chain.pem";

	// Mutate JsonConf
	await createLocation(JsonConf, { ...server, location: "/" });

	// Custom Locations
	if (server.locations.length) {
		await Promise.all(
			server.locations.map(async (location) => {
				await createLocation(JsonConf, location);
			})
		);
	}

	const config = parser.toConf(JsonConf);

	return config;
};

export default createConfig;
