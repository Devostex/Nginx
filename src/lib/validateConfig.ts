import chalk from "chalk";
import { pathExists } from "fs-extra";
import { resolve } from "path";
import { URL } from "url";
import { z, ZodIssueCode } from "zod";

import dnsLookup from "@utils/dnsLookup";
import log from "@utils/log";
import settings from "@utils/settings";

import { InputConfig, OutputConfig } from "@models/config";

const returnKeys = [
	"proxy_pass",
	"return",
	"redirect",
	"rewrite",
	"html",
	"static",
	"raw"
];
const optionalReturnKeys = ["raw"];

const literalSchema = z.union([z.string(), z.number(), z.boolean()]);
const literalOptArraySchema = z.union([
	literalSchema.transform((data) => [data]),
	literalSchema.array()
]);
export type Literals = z.output<typeof literalOptArraySchema>;

export const returnKeysFromOption = (
	options: Record<string, unknown>
): string[] => Object.keys(options).filter((key) => returnKeys.includes(key));

const urlSchema = z
	.string()
	.url()
	.refine((url) => ["http:", "https:"].includes(new URL(url).protocol), {
		message: "Invalid protocol"
	});

const authSchema = z
	.object({
		username: z
			.string()
			.refine((str) => !str.includes(":"), {
				message: `Username can't contain ":"`
			})
			.refine((str) => str.length, {
				message: "Username should not be empty"
			}),
		password: z.string().min(1)
	})
	.strict();

const proxyPassSchema = urlSchema
	.superRefine(async (url, ctx) => {
		if (settings.dontCheckDns) return;

		const { hostname } = new URL(url);
		const valid = await dnsLookup(hostname);
		if (!valid) {
			ctx.addIssue({
				message: chalk`DNS lookup failed for: {yellow ${hostname}}`,
				code: z.ZodIssueCode.custom
			});
		}
	})
	.transform((proxy_pass) => ({
		proxy_pass
	}));

const urlsOrUrlSchema = urlSchema
	.array()
	.or(urlSchema)
	.transform((data) => [data].flat());

const oneReturnRefinement =
	(allowNone = false) =>
	(options: Record<string, unknown>, { addIssue }: z.RefinementCtx): void => {
		const keys = returnKeysFromOption(options);

		if (!allowNone && keys.length == 0) {
			if (!("subdomains" in options) && !("locations" in options)) {
				return addIssue({
					message: chalk`Please specify at least one "return" type: {dim ${returnKeys.join(
						", "
					)}}`,
					code: "custom"
				});
			}
		}

		if (
			keys.filter((key) => !optionalReturnKeys.includes(key)).length > 1
		) {
			addIssue({
				message: chalk`Too many "return" types, found: {yellow ${keys.join(
					", "
				)}}, please use only one of them.`,
				code: "custom"
			});
		}
	};

// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
const pathNameSchema = (error: string) =>
	z
		.string()
		.transform((str) => resolve(str))
		.superRefine(async (path, ctx) => {
			if (!(await pathExists(path))) {
				ctx.addIssue({
					code: ZodIssueCode.custom,
					message: chalk`${error}: {dim ${path}}`
				});
			}
		});

const includePathNameSchema = pathNameSchema("Include path does not exists");

export const locationSchema = z
	.object({
		proxy_pass: urlSchema,
		websocket: z.boolean(),
		custom_css: urlsOrUrlSchema,
		custom_js: urlsOrUrlSchema,
		return: z.string().or(z.number()),
		headers: z.record(literalSchema),
		cors: z
			.boolean()
			.transform((bool) => (bool ? "*" : false))
			.or(
				urlSchema
					.transform((url) => new URL(url).origin)
					.or(
						z.string().refine((str) => str == "*", {
							message: 'use "*" as a wildcard'
						})
					)
			),
		redirect: z.string(),
		rewrite: z.string(),
		auth: authSchema.transform((auth) => [auth]).or(authSchema.array()),
		html: z.string(),
		static: pathNameSchema("Static path does not exist"),
		raw: z.record(literalOptArraySchema),
		include: includePathNameSchema
			.transform((string) => [string])
			.or(includePathNameSchema.array())
	})
	.partial()
	.strict();

const headersTransform = (
	options: z.infer<typeof locationSchema>
): z.infer<typeof locationSchema> => {
	const headers = {
		...options.headers,
		...("cors" in options &&
			options.cors && {
				"Access-Control-Allow-Origin": options.cors
			})
	};

	delete options.cors;

	return {
		...options,
		...(Object.keys(headers).length && { headers })
	};
};

export const locationsSchema = z
	.record(
		proxyPassSchema.or(
			locationSchema
				.superRefine(oneReturnRefinement(true))
				.transform(headersTransform)
		)
	)

	.transform((locations) =>
		Object.entries(locations ?? {}).map(([path, location]) => ({
			...location,
			location: path
		}))
	);

const subdomainSchema = locationSchema
	.extend({
		certbot_name: z.string(),
		disable_cert: z.boolean(),
		locations: locationsSchema
	})
	.partial();

// const recordRegex =
// 	(regex: RegExp, type: string) =>
// 	(record: Record<string, unknown>, ctx: z.RefinementCtx): void => {
// 		Object.keys(record).forEach((key) => {
// 			if (!regex.test(key)) {
// 				ctx.addIssue({
// 					code: z.ZodIssueCode.invalid_string,
// 					message: chalk`Not a valid ${type}: {dim ${key}}`,
// 					validation: "regex"
// 				});
// 			}
// 		});
// 	};

export const domainSchema = subdomainSchema
	.extend({
		subdomains: z.record(
			z.union([
				subdomainSchema
					.superRefine(oneReturnRefinement())
					.transform(headersTransform),
				proxyPassSchema
			])
		)
		// .superRefine(recordRegex(subdomainRegex, "subdomain"))
	})
	.partial();

const nginxSchema = z
	.object({
		log: z.string()
	})
	.partial()
	.strict();

export const configSchema = z
	.object({
		servers: z.record(
			z.union([
				domainSchema
					.superRefine(oneReturnRefinement())
					.transform(headersTransform),
				proxyPassSchema
			])
		),
		// .superRefine(recordRegex(domainRegex, "domain"))

		cloudflare: z.boolean(),
		nginx: nginxSchema
	})
	.partial()
	.strict();

const validateConfig = async (
	config: InputConfig | unknown
): Promise<OutputConfig | null> => {
	const result = await configSchema.spa(config);

	if (result.success) {
		return result.data;
	}

	log.invalidConfig(result.error.issues.length > 1);
	result.error.issues.forEach((issue) => {
		if (issue.code == z.ZodIssueCode.invalid_union) {
			const errors: string[] = issue.unionErrors.flatMap((error) =>
				error.issues.map((g) => g.message)
			);

			issue.message = `${Array.from(new Set(errors)).join(
				chalk` {gray or} `
			)}`;
		}
		log.configIssue(issue);
	});

	return null;
};

export default validateConfig;
