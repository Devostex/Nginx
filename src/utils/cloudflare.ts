import axios from "axios";
import { outputFile, pathExists } from "fs-extra";
import { join } from "path";

import log from "@utils/log";
import settings from "@utils/settings";
import store from "@utils/useStore";

import { IpsResponse } from "@models/cloudflare-ips";

enum Type {
	Cached = 0,
	Updated = 1,
	Unchanged = 2
}

type Data = {
	type: Type;
	ips: string[];
};

export const requestCloudflareIps = async (): Promise<Data> => {
	await store.read();

	log.updatingCloudflare();

	if (
		Date.now() - (store.data.cloudflare.updated ?? 0) <
		settings.cloudflareExpiry
	) {
		if (store.data.cloudflare.ips) {
			log.cloudflareCached();
			return { type: Type.Cached, ips: store.data.cloudflare.ips };
		}
	} else if (store.data.cloudflare.updated) {
		log.cloudflareExpired();
	}

	const started = Date.now();
	const res = await axios.get<{ result: IpsResponse }>(
		"https://api.cloudflare.com/client/v4/ips"
	);
	const took = Date.now() - started;

	const result = res.data.result;

	const etag = result.etag;
	const ips = result.ipv4_cidrs.concat(result.ipv6_cidrs);
	const changed = store.data.cloudflare.etag != etag;

	store.data.cloudflare.ips = ips;
	store.data.cloudflare.updated = Date.now();
	store.data.cloudflare.etag = etag;

	await store.write();

	if (store.data.cloudflare.ips && !changed) {
		log.cloudflareUnchanged(took);
		return { type: Type.Unchanged, ips };
	} else {
		log.cloudflareUpdated(took);
		return { type: Type.Updated, ips };
	}
};

export const updateCloudflareRealIp = async ({
	ips,
	type
}: Data): Promise<void> => {
	const cloudflareConfPath = join(
		settings.nginxConfigPath,
		"cloudflare.conf"
	);
	const configExists = await pathExists(cloudflareConfPath);

	if (!configExists || type == Type.Updated) {
		const realIpList =
			ips.map((ip) => `set_real_ip_from ${ip};`).join("\n") +
			"\n\nreal_ip_header CF-Connecting-IP;";

		await outputFile(cloudflareConfPath, realIpList);
		log.cloudflareRealIp();
	}
};
