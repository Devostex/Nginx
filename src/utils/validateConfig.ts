import { URL } from "url";
import { z } from "zod";

import log from "@utils/log";

import Config from "@models/config";

const urlSchema = z
	.string()
	.url()
	.refine((url) => ["http:", "https:"].includes(new URL(url).protocol), {
		message: "Invalid protocol"
	})
	.transform((url) => new URL(url).href);

const locationSchema = z
	.object({
		proxy_pass: urlSchema,
		websocket: z.boolean(),
		custom_css: z.array(urlSchema).or(urlSchema),
		custom_js: z.array(urlSchema).or(urlSchema)
	})
	.partial()
	.strict();

export const locationsSchema = z.record(z.union([locationSchema, urlSchema]));

const subdomainSchema = locationSchema
	.extend({
		locations: locationsSchema
	})
	.partial();

const serverSchema = subdomainSchema
	.extend({
		subdomains: z.record(z.union([subdomainSchema, urlSchema]))
	})
	.partial();

export const configSchema = z
	.object({
		servers: z.optional(z.record(serverSchema))
	})
	.strict();

const validateConfig = (config: Config): Config | null => {
	const result = configSchema.safeParse(config);
	if (!result.success) {
		result.error.issues.forEach((issue) => {
			if (issue.code == z.ZodIssueCode.invalid_union) {
				const expected = issue.unionErrors.flatMap((error) =>
					error.issues.map((g) =>
						"expected" in g ? g.expected : false
					)
				);

				issue.message = `Expected ${expected.join("/")}`;
			}
			log.configIssue(issue);
		});

		return null;
	} else {
		log.configValid();

		return result.data;
	}
};

export default validateConfig;
