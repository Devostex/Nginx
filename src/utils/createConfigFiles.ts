import { outputFile } from "fs-extra";
import { join } from "path";

import createConfig from "@lib/createConfig";
import { logger } from "@lib/logger";
import settings from "@utils/settings";

import { SimpleServer } from "@models/ParsedConfig";

/**
 * Creates config files from an array
 * The file is at `settings.nginxConfigPath/${i}-${server.filename}.conf`
 * @param servers An array of {@link SimpleServer}s
 * @param defaultUsername The default username used for the auth option
 */

export const createConfigFiles = (
	servers: SimpleServer[],
	defaultUsername: string | undefined
): Promise<void>[] => {
	return servers.map(async (server, i) => {
		const nginxConfig = await createConfig(server, defaultUsername);

		const fileName = join(
			settings.nginxConfigPath,
			`${i}-${server.filename}.conf`
		);

		await outputFile(fileName, nginxConfig);

		logger.configDone({ server });
	});
};
