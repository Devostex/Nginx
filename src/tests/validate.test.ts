import chalk from "chalk";
import { mocked } from "ts-jest/utils";

import validateConfig from "@lib/validateConfig";
import log from "@utils/log";

import fullErrorConfig from "@configs/full-error-config.json";
import fullValidConfig from "@configs/full-valid-config.json";
import invalidReturnTypesConfig from "@configs/invalid-return-types-config.json";

chalk.level = 0;

describe("Validate Config", () => {
	log.log = jest.fn().mockName("log.log");
	const mockedLog = mocked(log.log);

	test("It checks for an invalid config", async () => {
		expect(await validateConfig(fullErrorConfig)).toBe(null);

		expect(mockedLog.mock.calls.flat()).toMatchSnapshot();
	});

	test("It tests for invalid number of return types", async () => {
		expect(await validateConfig(invalidReturnTypesConfig)).toBe(null);

		expect(mockedLog.mock.calls.flat()).toMatchSnapshot();
	});

	test("It accepts a valid config", async () => {
		const value = await validateConfig(fullValidConfig);
		if (value == null) {
			console.log(mockedLog.mock.calls.flat());
		}
		expect(value).not.toBe(null);
		expect(value).toMatchSnapshot();
	});
});
