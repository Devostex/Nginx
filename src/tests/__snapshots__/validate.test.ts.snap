// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validate Config It accepts a valid config 1`] = `
{
  "nginx": {
    "log": "$time_local $remote_addr",
  },
  "servers": {
    "example.com": {
      "auth": [
        {
          "password": "password",
          "username": "username",
        },
      ],
      "certbot_name": "string",
      "custom_css": [
        "http://validurlstring",
      ],
      "custom_js": [
        "http://validurlarray",
      ],
      "headers": {
        "Access-Control-Allow-Origin": "*",
        "Content-Type": "text/html",
      },
      "html": "html string",
      "locations": [
        {
          "location": "indirect",
          "proxy_pass": "http://example.com",
        },
        {
          "location": "direct",
          "proxy_pass": "http://example.com",
        },
      ],
      "raw": {
        "custom": [
          "option",
        ],
      },
      "subdomains": {
        "no_return_allowed": {
          "locations": [
            {
              "html": "Return not needed in subdomain",
              "location": "/en",
            },
            {
              "headers": {
                "Access-Control-Allow-Origin": "https://return-not-needed-in-location",
              },
              "location": "/withCors",
            },
          ],
        },
        "redirects": {
          "redirect": "/",
        },
        "returns": {
          "return": "200 ok",
        },
        "rewrites": {
          "rewrite": "^ https://example.com",
        },
        "www": {
          "proxy_pass": "http://example.com",
        },
      },
      "websocket": true,
    },
  },
}
`;

exports[`Validate Config It checks for an invalid config 1`] = `
[
  "[ERROR] There was an issue with your config, the errors are listed below.",
  "[ERROR] Path: cloudflare",
  "[ERROR] Error: Expected boolean, received string",
  "[ERROR] Path: config",
  "[ERROR] Error: Unrecognized key(s) in object: 'unknown'",
  "[ERROR] Path: servers.example.com.proxy_pass",
  "[ERROR] Error: Invalid url",
  "[ERROR] Path: servers.example.com.headers",
  "[ERROR] Error: Expected object, received array",
  "[ERROR] Path: servers.example.com.redirect",
  "[ERROR] Error: Expected string, received array",
  "[ERROR] Path: servers.example.com.rewrite",
  "[ERROR] Error: Expected string, received array",
  "[ERROR] Path: servers.example.com.html",
  "[ERROR] Error: Expected string, received array",
  "[ERROR] Path: servers.example.com.certbot_name",
  "[ERROR] Error: Expected string, received array",
  "[ERROR] Path: servers.example.com.custom_css",
  "[ERROR] Error: Expected array, received object or Expected string, received object",
  "[ERROR] Path: servers.example.com.custom_js.0",
  "[ERROR] Error: Invalid url",
  "[ERROR] Path: servers.example.com.return",
  "[ERROR] Error: Expected string, received array or Expected number, received array",
  "[ERROR] Path: servers.example.com.auth.username",
  "[ERROR] Error: Username can't contain \\":\\"",
  "[ERROR] Path: servers.example.com.auth.password",
  "[ERROR] Error: Should be at least 1 characters",
  "[ERROR] Path: servers.example.com.auth",
  "[ERROR] Error: Unrecognized key(s) in object: 'unknown'",
  "[ERROR] Path: servers.example.com.locations.invalid",
  "[ERROR] Error: Invalid url",
  "[ERROR] Path: servers.example.com.cors",
  "[ERROR] Error: Invalid url or use \\"*\\" as a wildcard",
  "[ERROR] Path: servers.example.com.locations.extra.auth.username",
  "[ERROR] Error: Username should not be empty",
  "[ERROR] Path: servers.example.com.locations.extra.auth.password",
  "[ERROR] Error: Required",
  "[ERROR] Path: servers.example.com.subdomains.too_many",
  "[ERROR] Error: Too many \\"return\\" types, found: redirect, html, raw, please use only one of them.",
  "[ERROR] Path: servers.dns_error",
  "[ERROR] Error: DNS lookup failed for: not-existing",
]
`;

exports[`Validate Config It tests for invalid number of return types 1`] = `
[
  "[ERROR] There was an issue with your config, the errors are listed below.",
  "[ERROR] Path: servers.no returns",
  "[ERROR] Error: Please specify at least one \\"return\\" type: proxy_pass, return, redirect, rewrite, html, static, raw",
  "[ERROR] Path: servers.example.com",
  "[ERROR] Error: Too many \\"return\\" types, found: proxy_pass, redirect, rewrite, html, return, please use only one of them.",
]
`;
